src/client/Microsoft.Identity.Client/Credential/ManagedIdentityCredentialResponse.cs

```diff
@@ -0,0 +1,175 @@
+                    errorMessages.Add("TenantId is missing or empty. ");
+                }
+            }
+
+            // Check if any error messages were added
+            if (errorMessages.Any())
+            {
+                // Log an error message indicating the missing or insufficient fields
+                _requestContext.Logger.Error("[Managed Identity] " + string.Join(" ", errorMessages) +
+                    " and/or insufficient for authentication.");
+
+                // Throw an exception indicating that the CredentialResponse is invalid
+                MsalException exception = MsalServiceExceptionFactory.CreateManagedIdentityException(
+                    MsalError.ManagedIdentityRequestFailed,
+                    MsalErrorMessage.ManagedIdentityInvalidResponse,
+                    null,
+                    ManagedIdentitySource.Credential,
+                    null);
+
+                throw exception;
+            }
+        }
+
+        /// <summary>
+        /// Fetches a new managed identity credential from the IMDS endpoint.
+        /// </summary>
+        /// <param name="httpManager"></param>
+        /// <param name="cancellationToken"></param>
+        /// <returns></returns>
+        /// <exception cref="MsalManagedIdentityException"></exception>
+        private async Task<CredentialResponse> FetchFromServiceAsync(
+            IHttpManager httpManager,
+            CancellationToken cancellationToken)
+        {
+            _requestContext.Logger.Info("[Managed Identity] Fetching new managed identity credential from IMDS endpoint.");
+
+            OAuth2Client client = CreateClientRequest(httpManager);
+
+            CredentialResponse credentialResponse = await client
+                .GetCredentialResponseAsync(_uri, _requestContext)
+                .ConfigureAwait(false);
+
+            return credentialResponse;
+        }
+
+        /// <summary>
+        /// Creates an OAuth2 client request for fetching the managed identity credential.
+        /// </summary>
+        /// <param name="httpManager"></param>
+        /// <returns></returns>
+        private OAuth2Client CreateClientRequest(IHttpManager httpManager)
+        {
+            var client = new OAuth2Client(_requestContext.Logger, httpManager, null);
+
+            client.AddHeader("Metadata", "true");
+            client.AddHeader("x-ms-client-request-id", _requestContext.CorrelationId.ToString("D"));
+            string jsonPayload = GetCredentialPayload(_bindingCertificate);
+            client.AddBodyContent(new StringContent(jsonPayload, System.Text.Encoding.UTF8, "application/json"));
+
+            return client;
+        }
+
+        /// <summary>
+        /// Creates the payload for the managed identity credential request.
+        /// </summary>
+        /// <param name="x509Certificate2"></param>
+        /// <returns></returns>
+        private static string GetCredentialPayload(X509Certificate2 x509Certificate2)
```

Comment:
Code smell: x509certificate2 is a member of this class - it's better to use the member in this case. In other words - this method isn't static because it needs to read the class state.